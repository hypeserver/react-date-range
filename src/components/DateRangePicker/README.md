This component wraps **[DefinedRange](#definedrange)** and **[Calendar](#calendar)** components together, and extends all the props of them.

#### Example: 2 Month View

```jsx inside Markdown
import { addDays } from 'date-fns';
import { useState } from 'react';

const [state, setState] = useState([
  {
    startDate: new Date(),
    endDate: addDays(new Date(), 7),
    key: 'selection'
  }
]);

<DateRangePicker
  onChange={item => setState([item.selection])}
  showSelectionPreview={true}
  moveRangeOnFirstSelection={false}
  months={2}
  ranges={state}
  direction="horizontal"
/>;
```

#### Example: Vertical Infinite

```jsx inside Markdown
import { addDays } from 'date-fns';
import { useState } from 'react';

const [state, setState] = useState({
  selection: {
    startDate: new Date(),
    endDate: null,
    key: 'selection'
  },
  compare: {
    startDate: new Date(),
    endDate: addDays(new Date(), 3),
    key: 'compare'
  }
});

<DateRangePicker
  onChange={item => setState({ ...state, ...item })}
  months={1}
  minDate={addDays(new Date(), -300)}
  maxDate={addDays(new Date(), 900)}
  direction="vertical"
  scroll={{ enabled: true }}
  ranges={[state.selection, state.compare]}
/>;
```

#### Example: Multiple Range

```jsx inside Markdown
import { addDays } from 'date-fns';
import { useState } from 'react';

const [state, setState] = useState({
  selection1: {
    startDate: addDays(new Date(), 1),
    endDate: null,
    key: 'selection1'
  },
  selection2: {
    startDate: addDays(new Date(), 4),
    endDate: addDays(new Date(), 8),
    key: 'selection2'
  },
  selection3: {
    startDate: addDays(new Date(), 8),
    endDate: addDays(new Date(), 10),
    dayNumberColor: '#000000',
    key: 'selection3',
    autoFocus: false
  }
});

<DateRangePicker
  onChange={item => setState({ ...state, ...item })}
  ranges={[state.selection1, state.selection2, state.selection3]}
/>;
```

#### Example: Insert Aria-label

```jsx inside Markdown
import { addDays } from 'date-fns';
import { useState } from 'react';

const [state, setState] = useState([
  {
    startDate: addDays(new Date(), -6),
    endDate: new Date(),
    key: 'selection1'
  },
  {
    startDate: addDays(new Date(), 1),
    endDate: addDays(new Date(), 7),
    key: 'selection2'
  }
]);

<DateRangePicker
  onChange={item => setState([item.selection])}
  showSelectionPreview={true}
  moveRangeOnFirstSelection={false}
  months={2}
  ranges={state}
  direction="horizontal"
  ariaLabels={{
    dateInput: {
      selection1: { startDate: "start date input of selction 1", endDate: "end date input of selction 1" },
      selection2: { startDate: "start date input of selction 2", endDate: "end date input of selction 2" }
    },
    monthPicker: "month picker",
    yearPicker: "year picker",
    prevButton: "previous month button",
    nextButton: "next month button",
  }}
/>;
```