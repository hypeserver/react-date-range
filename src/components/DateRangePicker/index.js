import React, { Component } from 'react';
import PropTypes from 'prop-types';
import DateRange from '../DateRange';
import DefinedRange from '../DefinedRange';
import { findNextRangeIndex, generateStyles } from '../../utils';
import classnames from 'classnames';
import coreStyles from '../../styles';

class DateRangePicker extends Component {
  constructor(props) {
    super(props);
    this.state = {
      focusedRange: [findNextRangeIndex(props.ranges), 0],
    };
    this.styles = generateStyles([coreStyles, props.classNames]);
  }
  render() {
    const { focusedRange } = this.state;
    return (
      <div className={classnames(this.styles.dateRangePickerWrapper, this.props.className)}>
        <DefinedRange
          focusedRange={focusedRange}
          onPreviewChange={value =>
            this.dateRange.updatePreview(
              value ? this.dateRange.calcNewSelection(value, typeof value === 'string') : null
            )
          }
          {...this.props}
          range={this.props.ranges[focusedRange[0]]}
          className={undefined}
        />
        <DateRange
          onRangeFocusChange={focusedRange => this.setState({ focusedRange })}
          focusedRange={focusedRange}
          {...this.props}
          ref={t => (this.dateRange = t)}
          className={undefined}
        />
      </div>
    );
  }
}

DateRangePicker.defaultProps = {};

DateRangePicker.propTypes = {
  ...DateRange.propTypes,
  ...DefinedRange.propTypes,
  className: PropTypes.string,
};

export default DateRangePicker;
