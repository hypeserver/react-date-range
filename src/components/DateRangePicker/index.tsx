import React, { Component } from 'react';
import PropTypes from 'prop-types';
import DateRange, { DateRangeProps } from '../DateRange';
import DefinedRange, { DefinedRangeProps } from '../DefinedRange';
import { findNextRangeIndex, generateStyles } from '../../utils';
import classnames from 'classnames';
import coreStyles from '../../styles';
import { DateRange as DateRangeObject  } from '../../utilsTypes';
import { Preview, Styles } from '../DayCell/types';


interface DateRangePickerProps extends DateRangeProps, DefinedRangeProps  {
  classNames: string[]
  ranges: DateRangeObject[]
  className: string
  styles: Styles
}

interface DateRangePickerState {
  focusedRange: number[]
}


class DateRangePicker extends Component<DateRangePickerProps, DateRangePickerState> {
  static propTypes: any;
  static defaultProps: Styles;

  styles: Styles;
  // FIXME
  dateRange?: DateRange | null;

  constructor(props: DateRangePickerProps) {
    super(props);
    this.state = {
      focusedRange: [findNextRangeIndex(props.ranges), 0],
    };
    this.styles = generateStyles([coreStyles, props.classNames]);
  }
  render() {
    const { focusedRange } = this.state;
    return (
      <div className={classnames(this.styles.dateRangePickerWrapper, this.props.className)}>
        <DefinedRange
          {...this.props}
          focusedRange={focusedRange}
          onPreviewChange={value =>
            value && this.dateRange?.updatePreview(
              this.dateRange?.calcNewSelection(value, typeof value === 'string')
            )
          }
        />
        <DateRange
          {...this.props}
          onRangeFocusChange={focusedRange => this.setState({ focusedRange })}
          focusedRange={focusedRange}
          ref={t => (this.dateRange = t)}
        />
      </div>
    );
  }
}

DateRangePicker.propTypes = {
  ...DateRange.propTypes,
  ...DefinedRange.propTypes,
  className: PropTypes.string,
};

export default DateRangePicker;
