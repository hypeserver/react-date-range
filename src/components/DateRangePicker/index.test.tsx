import React from 'react';
import { mount, shallow } from 'enzyme';

import DateRangePicker from '.';

describe('DateRangePicker tests', () => {
  test('Should call "renderStaticRangeLabel" callback 0 times', () => {
    const renderStaticRangeLabel = jest.fn();
    mount(
      <DateRangePicker
        retainEndDateOnFirstSelection={false}
        weekStartsOn={1}
        onChange={item => {
          if (!item || item.selection === undefined)
            return;
          const selection = item.selection;
          renderStaticRangeLabel();
          console.log('selection start', selection.startDate);
          console.log('selection end', selection.endDate);
        }}
      />
    );
    expect(renderStaticRangeLabel).toHaveBeenCalledTimes(0);
  });

  test('Should render', () => {
    Date.now = jest.fn(() => 1482363367071);
    const wrapper = shallow(
      <DateRangePicker
        retainEndDateOnFirstSelection={false}
        weekStartsOn={0}
        onChange={item => {
          if (!item || item.selection === undefined)
            return;
          const selection = item.selection;
          console.log('selection start', selection.startDate);
          console.log('selection end', selection.endDate);
        }}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });
});