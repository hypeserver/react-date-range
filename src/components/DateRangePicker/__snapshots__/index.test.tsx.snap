// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateRangePicker tests Should render 1`] = `
<div
  className="rdrDateRangePickerWrapper"
>
  <DefinedRange
    className=""
    focusedRange={
      Array [
        0,
        0,
      ]
    }
    inputRanges={
      Array [
        Object {
          "getCurrentValue": [Function],
          "label": "days up to today",
          "range": [Function],
        },
        Object {
          "getCurrentValue": [Function],
          "label": "days starting today",
          "range": [Function],
        },
      ]
    }
    onChange={[Function]}
    onPreviewChange={[Function]}
    rangeColors={
      Array [
        "#3d91ff",
        "#3ecf8e",
        "#fed14c",
      ]
    }
    ranges={Array []}
    retainEndDateOnFirstSelection={false}
    staticRanges={
      Array [
        Object {
          "isSelected": [Function],
          "label": "Last Month",
          "range": [Function],
        },
        Object {
          "isSelected": [Function],
          "label": "Last Week",
          "range": [Function],
        },
        Object {
          "isSelected": [Function],
          "label": "Yesterday",
          "range": [Function],
        },
        Object {
          "isSelected": [Function],
          "label": "Today",
          "range": [Function],
        },
        Object {
          "isSelected": [Function],
          "label": "This Week",
          "range": [Function],
        },
        Object {
          "isSelected": [Function],
          "label": "This Month",
          "range": [Function],
        },
      ]
    }
    weekStartsOn={0}
  />
  <DateRange
    className=""
    classNames={Object {}}
    disabledDates={Array []}
    focusedRange={
      Array [
        0,
        0,
      ]
    }
    moveRangeOnFirstSelection={false}
    onChange={[Function]}
    onRangeFocusChange={[Function]}
    rangeColors={
      Array [
        "#3d91ff",
        "#3ecf8e",
        "#fed14c",
      ]
    }
    ranges={Array []}
    retainEndDateOnFirstSelection={false}
    weekStartsOn={0}
  />
</div>
`;
